### PR TITLE
pipes - Tamira and Irene - Scrabble 

### PR DESCRIPTION
**Pull request by tvojnar**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  We want to be able to score a word without creating an instance of that word  |
|   Describe an example of a test you wrote for a _nominal case_. |   When we tested that the self.score method returned the correct score for a word.	|
|   Describe an example of a test you wrote for an _edge case_. |  An edge case that we would have tested is to test that an ArgumentError was raised when a word was played in the play(word) method that was not included in the @tiles_in_hand array 	|
|   What was your final test coverage percentage?	|   95.43	|
|   Is there a specific piece of code you'd like feedback on?	|   Our play(word) method in our Player class. Also if there was a better way to make @tile instead of making it such a long array (in Tilebag class). Also if we were testing the right things (we know we didn't do an awesome job of  testing edge cases). 	|

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      array_of_words.each do |word|
+        word_score = Scrabble::Scoring.score(word)
+        if word_score > highest_score
+          highest_score = word_score
+          highest_score_word = word.upcase
+        elsif word_score == highest_score
+          if word.length == 7
+            highest_score_word = word.upcase
+          elsif word.length < highest_score_word.length
+            highest_score_word = word.upcase
```
Comment (droberts-sea) [inline lib/scoring.rb:31]: This tie-breaking logic gets pretty complex! It might be wise to break it out into a separate method, both to clean up this code and so that the logic can be tested independently.

Also, I'm not sure this logic is 100% correct. For example, if you have a list with multiple 7-letter words, it looks like it will take the last, not the first.

File: specs/scoring_spec.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      Scrabble::Scoring.score("HELLO").must_be_kind_of Integer
+    end #it "will return an integer score"
+
+    it "will return the right score for cat" do
+      Scrabble::Scoring.score("CAT").must_equal 5
+    end #it "will return the right score for a word" do
+
+    it "will return the right score for elephant" do
+      Scrabble::Scoring.score("turkey").must_equal 13
+    end #it "will return the right score for elephant" do
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:39]: This test does not match the name.

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    it "will add 50 to a 7 letter word" do
+      Scrabble::Scoring.score("CATTTTT").must_equal 59
+    end
+
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:44]: What about an empty word, or an 8-letter word?

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    end #it "if equal, should pick the word with 7 letters" do
+
+    it "if equal, it should pick the word with least number of letters" do
+      Scrabble::Scoring.highest_score_from_array(["dog", "train"]).must_equal "DOG"
+    end #it "if equal, should pick the word with 7 letters" do
+
+    it "if equal in score and length, it should pick the first word in the list" do
+      Scrabble::Scoring.highest_score_from_array(["dog", "god"]).must_equal "DOG"
+    end #it "if equal in score and length, it should pick the first word in the list" do
+
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:68]: You're missing a few cases here. What happens when this method is called on:
- An empty array
- An array with one element
- An array with a 7-letter and 6-letter word with the same score (e.g. `ZZZZZZ` and `AAAAAAW`)?
- An array with two 7-letter words with the same score

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    end #def highest_score_word(word_array)
+
+    def highest_word_score
+      # Returns the highest scoring total as an integer
+      Scrabble::Scoring.score(highest_scoring_word)
+    end
+
+    def won?
+      if @total_score > 100
+        return true
```
Comment (droberts-sea) [inline lib/player.rb:57]: Since `<` always returns either `true` or `false` already, this could be shortened to:

```ruby
def won?
  return @total_score > 100
end
```

File: lib/tilebag.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+module Scrabble
+  class Tilebag
+    attr_reader :tiles
+
+    def initialize
+      @tiles = ['A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'D', 'D', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'E', 'F', 'F', 'G', 'G', 'G', 'H', 'H', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'J', 'K', 'L', 'L', 'L', 'L', 'M', 'M', 'N', 'N', 'N', 'N', 'N', 'N', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'P', 'P', 'Q', 'R', 'R', 'R', 'R', 'R', 'R', 'S', 'S', 'S', 'S', 'T', 'T', 'T', 'T', 'T', 'T', 'U', 'U', 'U', 'U', 'V', 'V', 'W', 'W', 'X', 'Y', 'Y', 'Z']
+    end
+
```
Comment (droberts-sea) [inline lib/tilebag.rb:8]: It might make sense here to store the counts in a hash, mapping letter to number of tiles. Then in initialize, you could iterate through that hash to populate the array. This is more work but it would be much easier to read.

File: specs/tile_bag_spec.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      picked_tiles.length.must_equal 98
+    end
+
+    it "should remove each tile from the @tile" do
+      starting_tiles = @tilebag.tiles.length
+      num = 10
+      @tilebag.draw_tiles(num)
+      @tilebag.tiles.length.must_equal starting_tiles - num
+    end
+
```
Comment (droberts-sea) [inline specs/tile_bag_spec.rb:41]: You're missing a couple of interesting edge cases here:
- What happens if you attempt to draw 0 tiles?
- What happens if you attempt to draw the last tile from the bag?
- What happens when you attempt to draw more tiles than are in the bag?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  yes  |
|   Both Teammates contributed to the codebase	|  yes  |
|   Regular Commits with meaningful commit messages  |  yes  |
|  **Scoring class**	|   	|
|  `score` method	|  yes  |
|   Uses appropriate data structure to store the letter score	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  some - see inline  |
|  `highest_scoring_word` method	|  yes  |
|   Appropriately handles edge cases	|  mostly - see inline  |
|   Tests `highest_scoring_word` edge cases (ties, empty array)	|  some - see inline  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  yes  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  yes  |
|  Returns `true` or `false` from `won?` method |  yes  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  no - what happens if the player has already won? Does their score or word list change?  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  yes  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  yes  |
| `tiles_remaining` method returns a count	|  yes  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|  some - see inline  |

Aside from the bug in `highest_scoring_word`, the product code for this project looks pretty good.

However, your test coverage has some gaps, which I've tried to outline below. The general pattern seems to be that while you have good coverage of nominal cases (cases in which your program behaves as expected), you miss failure or error cases. For example, in `TileBag#draw_tiles`, you don't consider what happens if you attempt to draw more tiles than are in the bag. Particularly interesting to me is the place where the input changes from OK to problematic. To use the same example, this would be drawing the last tile in the bag. You can look at your code and say "of course that would work", but until you test it you can't know for sure.

Becoming a master tester is not easy, and this is the first project where you've had to build tests around complex logic, so don't worry if it's not yet where we want it to be. We'll have plenty of time to practice going forward. Do make sure to spend time thinking about it for Hotel, particularly around how ranges of dates interact, and let one of us know if you want us to check your work.
